Scheduler refactoring

andyross opened the pull request:
Big pull request which in theory shouldn't change any behavior (only done sanitycheck over tests/kernel, and only on x86 so far though!).

The net effect is to eliminate most of the use of our fine-grained scheduler tooling in favor of coarser functions with clear semantics, easier replaceability, and (most importantly) clean synchronization expectations instead of "must be called with irq_lock held" all over the place.  Almost everything comes down to:

1. Pend the current thread and context switch to whatever is next
2. Unpend one thread from a wait_q and make it runnable
3. Unpend a specific thread from wherever it is sleeping
4. Check the scheduling context and context switch ("_reschedule") if needed

Commit:
kernel: Allow pending w/o wait_q for scheduler API cleanup

The mailbox code was written to use the _remove_thread_from_ready_q()
API directly, which would be good to get out of the scheduler internal
API.  What it really wanted to do is to mark a thread "PENDING"
without actually adding it to a wait queue, which is sane enough (the
message stores the "thread to wake up on receipt" handle).

So allow that naturally in the _pend_thread() API by passing a NULL
wait_q.  Really a wait_q needn't be the only way a thread can block.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Scheduler refactoring: use _reschedule_*() always

There was a somewhat promiscuous pattern in the kernel where IPC
mechanisms would do something that might effect the current thread
choice, then check _must_switch_threads() (or occasionally
__must_switch_threads -- don't ask, the distinction is being replaced
by real English words), sometimes _is_in_isr() (but not always, even
in contexts where that looks like it would be a mistake), and then
call _Swap() if everything is OK, otherwise releasing the irq_lock().
Sometimes this was done directly, sometimes via the inverted test,
sometimes (poll, heh) by doing the test when the thread state was
modified and then needlessly passing the result up the call stack to
the point of the _Swap().

And some places were just calling _reschedule_threads(), which did all
this already.

Unify all this madness.  The old _reschedule_threads() function has
split into two variants: _reschedule_yield() and
_reschedule_noyield().  The latter is the "normal" one that respects
the cooperative priority of the current thread (i.e. it won't switch
out even if there is a higher priority thread ready -- the current
thread has to pend itself first), the former is used in the handful of
places where code was doing a swap unconditionally, just to preserve
precise behavior across the refactor.  I'm not at all convinced it
should exist...

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Refactor, unifying _pend_current_thread() + _Swap() idiom

Everywhere the current thread is pended, the code is going to have to
do a _Swap() soon afterward, yet the scheduler API exposed these as
separate steps.  Unify this pattern everywhere it appears, which saves
some code bytes and gets _Swap() out of the general scheduler API at
zero cost.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: include cleanup

Recent changes have eliminated most use of _Swap() in favor of higher
level scheduler abstractions.  We can remove the header too.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Clean up _unpend_thread() API

Almost everywhere this was called, it was immediately followed by
_abort_thread_timeout(), for obvious reasons.  The only exceptions
were in timeout and k_timer expiration (unifying these two would be
another good cleanup), which are peripheral parts of the scheduler and
can plausibly use a more "internal" API.

So make the common case the default, and expose the old behavior as
_unpend_thread_no_timeout().  (Along with identical changes for
_unpend_first_thread) Saves code bytes and simplifies scheduler
surface area for future synchronization work.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Further unify _reschedule APIs

Now that other work has eliminated the two cases where we had to do a
reschedule "but yield even if we are cooperative", we can squash both
down to a single _reschedule() function which does almost exactly what
legacy _Swap() did, but wrapped as a proper scheduler API.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Deprecate k_thread_cancel() API

The only difference between this call and k_thread_abort() (beyond
some minor performance deltas) is that "cancel" will act as a noop in
cases where the thread has begun execution and will return an error.
"Abort" always succeeds, of course.  That is inherently racy when used
as a "stop the thread" API: there's no way in general (or at all in
SMP situations) to know that you're calling this function "early
enough" to catch the thread before it starts.

Effectively, all k_thread_cancel() gives you that k_thread_abort()
doesn't is an indication about whether or not a thread has started.
There are many other ways to get that information that don't require
dangerous kernel APIs.

Deprecate this function.  Zephyr's own code never used it except for
its own unit test.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lpereira commented on include/kernel.h:
```diff
@@ -716,8 +716,10 @@ __syscall k_tid_t k_current_get(void);
+ * @deprecated This API is deprecated.  Use k_thread_abort().
  */
-__syscall int k_thread_cancel(k_tid_t thread);
+ __deprecated __syscall int k_thread_cancel(k_tid_t thread);
```
Shouldn't __syscall be the first thing in this line?

andrewboie replied:
I think this is actually the right way to do this. However, please check the generated syscall_dispatch.c to ensure it is there.

andyross replied:
Actually, it doesn't work if you put __syscall first.  The script appears to suck "__deprecated" into the function siganture of the implementation, which then spews warnings when referenced.  I think.  All I know is that I had it that way first, got warnings that killed the build, then swapped it to fix it. :)

tyrantlizard77 reviewed (approved):
Looks great.

I notice some space/tab issues but checkpatch will barf on that until it's fixed, so +1 from me.

Commit:
pthread: Respect cooperative thread schedulign in condition variable

Originally, pthread_cond_signal() was written to yield even in
circumstances where the current thread is at a cooperative priority
and would not expect to be context-switched out until it blocks.  This
makes sense, as in most cases you want the newly signaled thread to
get a chance to run as soon as possible.

On further reflection (and also because it complicates the scheduler),
I think that's wrong.  The point to cooperative scheduling is that it
allows the cooperative code to make synchronization assumptions about
exactly when it might yield to other threads, and having arbitrary
APIs be "preemption points" like this complicates that analysis
significantly.

Use _reschedule() like other code does.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lpereira commented on kernel/mutex.c:
```diff
@@ -247,6 +239,9 @@ void _impl_k_mutex_unlock(struct k_mutex *mutex)
+		_ready_thread(new_owner);
+		irq_unlock(key);
 	} else {
 		irq_unlock(key);
```
Maybe you can move the call to irq_unlock() outside this if/else block, right before k_sched_unlock()?

(I'm not sure if there's anything that assumes that mutex->owner will be set to NULL with the interrupts unlocked, though.)

andyross replied:
Yeah, that's cleaner. Honestly I think putting the owner assignment outside the lock is a race to begin with. And in fact the block here is tested by if(new_owner) and mutex->owner gets set to that, so we can move the asssignment to before the block and the unlock to after and it's cleaner still. Thanks.

lpereira reviewed (approved):
Just a minor nit in addition to the indentation issue that @andrewboie commented about, but otherwise LGTM

Commit:
kernel/mutex: Clean up k_mutex_unlock()

Recent changes to the scheduler API means we can simplify this
further: move the assignment to mutex->owner outside the if(), which
removes the need to have an else clause (which just set that field to
NULL when the new_owner was already NULL); and we can likewise move
the irq_unlock() outside the block.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
@lpereira noticed that I somehow had "restored" an ancient gerrit git hook and was emitting patches with Change-Id: headers.  Fixed.

andyross commented:
Refresh this, having founds some spots where some "obvious cleanup" was not a 1:1 refactoring.  Shame on me.

Commit:
kernel, posix: Move ready_one_thread() to scheduler

The POSIX layer had a simple ready_one_thread() utility.  Move this to
the scheduler API (with a prepended underscore -- it's an internal
API) so that it can be synchronized along with the rest of the
scheduler.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lpereira commented:
What has changed since the last time this was reviewed?

andyross commented:
Rebased
